admin/menu/class-base-menu.php

```diff
@@ -0,0 +1,192 @@
+<?php
+/**
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Admin\Menu
+ */
+
+/**
+ * Admin menu base class.
+ */
+abstract class WPSEO_Base_Menu implements WPSEO_WordPress_Integration {
+
+	/** @var WPSEO_Menu Menu */
+	protected $menu;
+
+	/**
+	 * Constructs the Admin Menu.
+	 *
+	 * @param WPSEO_Menu $menu Menu to use.
+	 */
+	public function __construct( WPSEO_Menu $menu ) {
+		$this->menu = $menu;
+	}
+
+	/**
+	 * Registers all hooks to WordPress.
+	 *
+	 * @return void
+	 */
+	public abstract function register_hooks();
+
+	/**
+	 * Returns the list of registered submenu pages.
+	 *
+	 * @return array List of registered submenu pages.
+	 */
+	public abstract function get_submenu_pages();
+
+	/**
+	 * Creates a submenu formatted array.
+	 *
+	 * @param string     $page_title Page title to use.
+	 * @param string     $page_slug  Page slug to use.
+	 * @param callable   $callback   Optional. Callback which handles the page request.
+	 * @param callable[] $hook       Optional. Hook to trigger when the page is registered.
+	 *
+	 * @return array Formatted submenu.
+	 */
+	protected function get_submenu_page( $page_title, $page_slug, $callback = null, $hook = null ) {
+		if ( $callback === null ) {
+			$callback = $this->get_admin_page_callback();
+		}
+
+		return array(
+			$this->get_page_identifier(),
+			'',
+			$page_title,
+			$this->get_manage_capability(),
+			$page_slug,
+			$callback,
+			$hook,
+		);
+	}
+
+	/**
+	 * Registers submenu pages as menu pages.
+	 *
+	 * @param array $submenu_pages List of submenu pages.
+	 */
+	protected function register_menu_pages( $submenu_pages ) {
+		if ( ! is_array( $submenu_pages ) || $submenu_pages === array() ) {
+			return;
+		}
+
+		// Loop through submenu pages and add them.
+		foreach ( $submenu_pages as $submenu_page ) {
```

Comment:
What about `array_walk` here? That calls a method: `add_menu_page` which contains the code from the foreach?